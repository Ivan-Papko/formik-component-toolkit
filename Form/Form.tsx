import { Formik } from 'formik'

import { FormikMultiSelect } from './FormikMultiSelect/FormikMultiSelect'
import { FormikAutocomplete } from './FormikAutocomplete'
import { FormikCheckbox } from './FormikCheckbox'
import { FormikCopyInput } from './FormikCopyInput'
import { FormikCountryField } from './FormikCountryField'
import { FormikDatePicker } from './FormikDatePicker'
import { FormikForm } from './FormikForm'
import { FormikPasswordField } from './FormikPasswordField'
import { FormikRegionField } from './FormikRegionField'
import { FormikSelect } from './FormikSelect'
import { FormikSubmit } from './FormikSubmit'
import { FormikSwitch } from './FormikSwitch'
import {
  type IFormikTextFieldProps,
  FormikTextField,
  FormikTextFieldWithRef,
} from './FormikTextField'
import { FormikTimePicker } from './FormikTimePicker'
import { FormikValidatingInput } from './FormikValidatingInput'
import { FormikWrapper } from './FormikWrapper'

/**
 * @deprecated Use `HookForm` instead.
 */
const Form = {
  Context: Formik,
  Form: FormikForm,
  Wrapper: FormikWrapper,
  TextField: FormikTextField,
  TextFieldWithRef: FormikTextFieldWithRef,
  PasswordField: FormikPasswordField,
  Checkbox: FormikCheckbox,
  ValidatingInput: FormikValidatingInput,
  CountryField: FormikCountryField,
  /**
   * @deprecated
   */
  RegionField: FormikRegionField,
  Submit: FormikSubmit,
  Switch: FormikSwitch,
  MultiSelect: FormikMultiSelect,
  Select: FormikSelect,
  Autocomplete: FormikAutocomplete,
  CopyInput: FormikCopyInput,
  TimePicker: FormikTimePicker,
  DatePicker: FormikDatePicker,
}

export { Form, type IFormikTextFieldProps }
export default Form
