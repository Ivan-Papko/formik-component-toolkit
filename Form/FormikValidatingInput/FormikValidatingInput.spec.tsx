import type { Dispatch, SetStateAction } from 'react'
import { waitFor } from '@testing-library/react'

import { render } from '@cloudbees/shared-test-utils/render'

import { Form } from '../Form'

import type { FormikValidationInputComponentState } from './FormikValidatingInput'

interface FactoryProps {
  formikId: string
  validationCallback: jest.Mock<Promise<unknown>>
  label: string
  validationErrorMessage: string
  hintMessage: string
  successMessage: (message: string) => string
  placeholderMessage: string
  componentStateSetter?: Dispatch<SetStateAction<FormikValidationInputComponentState>>
  defaultValue?: string
  onChange?: (input: string) => void
}

jest.mock('@mui/material/utils/debounce', () => jest.fn(fn => fn))

const mockValidationCallback = jest.fn(input => {
  return new Promise(resolve => {
    setTimeout(() => {
      if (input === 'valid') {
        resolve(true)
      } else {
        resolve(false)
      }
    }, 500)
  })
})

const defaultProps = {
  formikId: 'organizations',
  validationCallback: mockValidationCallback,
  label: 'Validating Input',
  validationErrorMessage: 'Invalid input',
  hintMessage: 'HintMessage',
  successMessage: (message: string) => message,
  placeholderMessage: 'Enter text',
}

const factory = ({
  formikId,
  validationCallback,
  label,
  validationErrorMessage,
  hintMessage,
  successMessage,
  placeholderMessage,
  ...props
}: FactoryProps) =>
  render(
    <Form.Context initialValues={{}} onSubmit={jest.fn()}>
      <Form.Form>
        <Form.Wrapper>
          <Form.ValidatingInput
            formikId={formikId}
            validationCallback={validationCallback}
            label={label}
            validationErrorMessage={validationErrorMessage}
            hintMessage={hintMessage}
            successMessage={successMessage}
            placeholderMessage={placeholderMessage}
            {...props}
          />
        </Form.Wrapper>
      </Form.Form>
    </Form.Context>,
  )

const mockComponentStateSetter = jest.fn()

describe('FormikValidatingInput', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.runOnlyPendingTimers()
    jest.useRealTimers()
    mockValidationCallback.mockClear()
    mockComponentStateSetter.mockClear()
  })
  it('should render a ValidatingInput component', async () => {
    const { baseElement } = factory({ ...defaultProps, defaultValue: 'My Licence' })
    expect(baseElement).toMatchSnapshot()
  })

  it('should validate input after debounce period', async () => {
    const { screen, setup } = factory({
      ...defaultProps,
      componentStateSetter: mockComponentStateSetter,
    })

    const user = setup()

    const input = screen.getByPlaceholderText('Enter text')
    await user.type(input, 'invalid')

    jest.advanceTimersByTime(600)

    await waitFor(() => expect(mockValidationCallback).toHaveBeenCalledWith('invalid'))
    expect(mockComponentStateSetter).toHaveBeenCalledWith({
      isValid: false,
      returnValue: false,
      currentInputValue: 'invalid',
    })
  })

  it('should show success message for valid input', async () => {
    const { screen, setup } = factory({
      ...defaultProps,
      componentStateSetter: mockComponentStateSetter,
    })
    const user = setup()
    const input = screen.getByPlaceholderText('Enter text')

    await user.type(input, 'valid')

    jest.advanceTimersByTime(600)

    await waitFor(() => expect(mockValidationCallback).toHaveBeenCalledWith('valid'))

    expect(mockComponentStateSetter).toHaveBeenCalledWith({
      isValid: true,
      returnValue: true,
      currentInputValue: 'valid',
    })
  })
})
