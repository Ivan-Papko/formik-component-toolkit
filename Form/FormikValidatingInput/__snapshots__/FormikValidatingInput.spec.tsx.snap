// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormikValidatingInput should render a ValidatingInput component 1`] = `
@keyframes animation-0 {
  0% {
    -webkit-transform: rotate(0deg);
    -moz-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  100% {
    -webkit-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-1 {
  0% {
    stroke-dasharray: 1px,200px;
    stroke-dashoffset: 0;
  }

  50% {
    stroke-dasharray: 100px,200px;
    stroke-dashoffset: -15px;
  }

  100% {
    stroke-dasharray: 100px,200px;
    stroke-dashoffset: -125px;
  }
}

.emotion-0 {
  margin-top: 24px;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 16px;
  width: 100%;
}

.emotion-1 {
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  min-width: 0;
  padding: 0;
  margin: 0;
  border: 0;
  vertical-align: top;
}

.emotion-2 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.4375em;
  letter-spacing: 0.00938em;
  color: rgba(0, 0, 0, 0.87);
  box-sizing: border-box;
  position: relative;
  cursor: text;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  padding-right: 14px;
}

.emotion-2.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
  cursor: default;
}

.emotion-2:hover .MuiOutlinedInput-notchedOutline {
  border-color: rgba(0, 0, 0, 0.87);
}

@media (hover: none) {
  .emotion-2:hover .MuiOutlinedInput-notchedOutline {
    border-color: rgba(0, 0, 0, 0.23);
  }
}

.emotion-2.Mui-focused .MuiOutlinedInput-notchedOutline {
  border-color: #1976d2;
  border-width: 2px;
}

.emotion-2.Mui-error .MuiOutlinedInput-notchedOutline {
  border-color: #d32f2f;
}

.emotion-2.Mui-disabled .MuiOutlinedInput-notchedOutline {
  border-color: rgba(0, 0, 0, 0.26);
}

.emotion-3 {
  font: inherit;
  letter-spacing: inherit;
  color: currentColor;
  padding: 16.5px 14px;
  border: 0;
  box-sizing: content-box;
  background: none;
  height: 1.4375em;
  margin: 0;
  -webkit-tap-highlight-color: transparent;
  display: block;
  min-width: 0;
  width: 100%;
  -webkit-animation-name: mui-auto-fill-cancel;
  animation-name: mui-auto-fill-cancel;
  -webkit-animation-duration: 10ms;
  animation-duration: 10ms;
  padding-right: 0;
}

.emotion-3::-webkit-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-3::-moz-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-3:-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-3::-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-3:focus {
  outline: 0;
}

.emotion-3:invalid {
  box-shadow: none;
}

.emotion-3::-webkit-search-decoration {
  -webkit-appearance: none;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3::-webkit-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3::-moz-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3:-ms-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3::-ms-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3:focus::-webkit-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3:focus::-moz-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3:focus:-ms-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3:focus::-ms-input-placeholder {
  opacity: 0.42;
}

.emotion-3.Mui-disabled {
  opacity: 1;
  -webkit-text-fill-color: rgba(0, 0, 0, 0.38);
}

.emotion-3:-webkit-autofill {
  -webkit-animation-duration: 5000s;
  animation-duration: 5000s;
  -webkit-animation-name: mui-auto-fill;
  animation-name: mui-auto-fill;
}

.emotion-3:-webkit-autofill {
  border-radius: inherit;
}

.emotion-4 {
  display: flex;
  height: 0.01em;
  max-height: 2em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  white-space: nowrap;
  color: rgba(0, 0, 0, 0.54);
  margin-left: 8px;
}

.emotion-5 {
  display: inline-block;
  color: #1976d2;
  -webkit-animation: animation-0 1.4s linear infinite;
  animation: animation-0 1.4s linear infinite;
}

.emotion-6 {
  display: block;
}

.emotion-7 {
  stroke: currentColor;
  stroke-dasharray: 80px,200px;
  stroke-dashoffset: 0;
  -webkit-animation: animation-1 1.4s ease-in-out infinite;
  animation: animation-1 1.4s ease-in-out infinite;
}

.emotion-8 {
  text-align: left;
  position: absolute;
  bottom: 0;
  right: 0;
  top: -5px;
  left: 0;
  margin: 0;
  padding: 0 8px;
  pointer-events: none;
  border-radius: inherit;
  border-style: solid;
  border-width: 1px;
  overflow: hidden;
  min-width: 0%;
  border-color: rgba(0, 0, 0, 0.23);
}

.emotion-9 {
  float: unset;
  width: auto;
  overflow: hidden;
  padding: 0;
  line-height: 11px;
  -webkit-transition: width 150ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: width 150ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
}

.emotion-10 {
  color: rgba(0, 0, 0, 0.6);
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.75rem;
  line-height: 1.66;
  letter-spacing: 0.03333em;
  text-align: left;
  margin-top: 3px;
  margin-right: 14px;
  margin-bottom: 0;
  margin-left: 14px;
}

.emotion-10.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
}

.emotion-10.Mui-error {
  color: #d32f2f;
}

<body>
  <div>
    <form
      action="#"
    >
      <div
        class="MuiBox-root emotion-0"
      >
        <div
          class="MuiFormControl-root MuiTextField-root emotion-1"
          data-testid="validation-input-contents"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd emotion-2"
          >
            <input
              aria-describedby="organizations-helper-text"
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd emotion-3"
              data-testid="organizations-input"
              id="organizations"
              name="organizations"
              placeholder="Enter text"
              type="text"
              value="My Licence"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium emotion-4"
            >
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary emotion-5"
                data-testid="licenses-page-progress"
                role="progressbar"
                style="width: 14px; height: 14px;"
              >
                <svg
                  class="MuiCircularProgress-svg emotion-6"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate emotion-7"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline emotion-8"
            >
              <legend
                class="emotion-9"
              >
                <span
                  class="notranslate"
                >
                  â€‹
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled emotion-10"
            id="organizations-helper-text"
          />
        </div>
        <small
          class="text-muted mb-3 form-text"
        >
          HintMessage
        </small>
      </div>
    </form>
  </div>
</body>
`;
